### PR TITLE
Prevent go-cmp from being imported in production code

### PR DESCRIPTION
**Pull request by rosstimothy**
Adds a new depguard rule to limit the use of go-cmp to test code. There are a few notable exceptions for legacy code that relies on go-cmp which should find other suitable means for comparisions. However, in the interest in limiting new code from introducing more production depenency on go-cmp, the existing violations are allowed for the time being and should be refactored and removed from the list in the future.

### CODE REVIEW

**Review by hugoShaka — APPROVED** [review]

File: .golangci.yml [hunk 1]
```diff
@@ -185,6 +185,36 @@ linters-settings:
...
             desc: '"vnet/daemon" requires CGO'
+      # Prevent importing go-cmp into production code. From the go-cmp docs:
+      # > It is intended to only be used in tests, as performance is not a goal
+      # > and it may panic if it cannot compare the values. Its propensity towards
+      # > panicking means that its unsuitable for production environments where a
+      # > spurious panic may be fatal.
+      go-cmp:
+        files:
+          # Tests can do anything
+          - '!$test'
```
Comment (codingllama) [inline .golangci.yml:196]: Should this say something like `!**/*_test.go` or is "!$test" a magical word already?

File: .golangci.yml [hunk 2]
```diff
@@ -185,6 +185,36 @@ linters-settings:
...
+          # Various test helpers defined outside _test.go files are allowed
+          - '!**/integration/helpers/**'
+          - '!**/integrations/operator/controllers/resources/testlib/**'
+          - '!**/lib/auth/test/**'
+          - '!**/lib/services/suite/**'
+          # Non-compliant legacy code. These should be converted to compare by another mechanism
+          # and be removed from this list in the future. Use caution before adding any additional
+          # exclusions to this list.
+          - '!**/e/lib/accesslist/equal.go'
```
Comment (codingllama) [inline .golangci.yml:205]: Do these need the "**" at the start or can we target more specifically?

File: .golangci.yml [hunk 3]
```diff
@@ -185,6 +185,36 @@ linters-settings:
...
+          - '!**/e/lib/auth/saml.go'
+          - '!**lib/services/authority.go'
+          - '!**lib/services/compare.go'
+          - '!**/lib/services/local/access_list.go'
+          - '!**/lib/services/local/users.go'
```
Comment (codingllama) [inline .golangci.yml:210]: 😱

**Review by codingllama — COMMENTED** [review]
Many thanks! Added a couple of questions but LGTM.

**Review by codingllama — APPROVED** [review]

**Pull request merged**